### PR TITLE
Feature/more errors

### PR DESCRIPTION
**Pull request by pavlo-hilei**
- Add tests for all possible errors
- Fix normalizer working with string parameters in attributes
- Add loop structure verification
- Move caret at the first non-white character in case of regular error in the modified line
- Fix check between outer and inner for @shared and @exclusive

### CODE REVIEW

File: include/oklt/util/string_utils.h [hunk 1]
```diff
@@ -22,6 +22,7 @@ std::string pointerToStr(const void* ptr);
                                                   const std::string_view& sep,
                                                   int maxN = -1,
                                                   bool keepEmpty = true);
+std::string replace(std::string_view str, std::string_view from, std::string_view to);
```
Comment (kchabSS) [inline include/oklt/util/string_utils.h:25]: Add `[[nodiscard]]` please.

File: lib/attributes/utils/common.cpp [hunk 2]
```diff
@@ -0,0 +1,20 @@
+#include "attributes/utils/common.h"
+
+namespace oklt {
+bool isLastOuter(OklLoopInfo* loop) {
```
Comment (kchabSS) [inline lib/attributes/utils/common.cpp:4]: It is better to be a part of `OklLoopInfo`

File: lib/attributes/utils/kernel_utils.h [hunk 3]
```diff
@@ -9,7 +9,8 @@ class AttributedStmt;
 }  // namespace clang
 
 namespace oklt {
-tl::expected<void, Error> verifyLoops(OklSemaCtx::ParsedKernelInfo& kernelInfo);
+tl::expected<void, Error> verifyLoops(OklSemaCtx::ParsedKernelInfo& kernelInfo,
```
Comment (kchabSS) [inline lib/attributes/utils/kernel_utils.h:12]: Please move `SessionStage&` to be first parameter.

File: lib/attributes/utils/serial_subset/shared.cpp [hunk 4]
```diff
@@ -1,7 +1,7 @@
+#include "attributes/utils/common.h"
```
Comment (kchabSS) [inline lib/attributes/utils/serial_subset/shared.cpp:1]: Please sort global/local includes properly.

File: lib/util/string_utils.cpp [hunk 5]
```diff
@@ -85,6 +83,19 @@ std::vector<std::string_view> split(const std::string_view& str,
     return ret;
 }
 
+std::string replace(std::string_view str, std::string_view from, std::string_view to) {
```
Comment (kchabSS) [inline lib/util/string_utils.cpp:86]: `StringRef` already has this mechanism. Please use it if possible.

File: tests/functional/configs/test_suite_transpiler/backends/cuda/barrier.json [hunk 6]
```diff
@@ -9,5 +9,16 @@
...
+        "action_config": {
+            "backend": "cuda",
+            "source": "transpiler/backends/cuda/barrier/barrier_warp.cpp",
+            "includes": [],
+            "defs": [],
+            "launcher": ""
+        },
+        "reference": "transpiler/backends/cuda/barrier/barrier_warp_ref.cpp"
      }
 ]
```
Comment (kchabSS) [inline tests/functional/configs/test_suite_transpiler/backends/cuda/barrier.json:24]: EOL at EOF

File: tests/functional/configs/test_suite_transpiler/common/errors/atomic.json [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+        "action_config": {
+            "backend": "cuda",
+            "source": "transpiler/common/errors/atomic/is_not_assignable.cpp",
+            "includes": [],
+            "defs": [],
+            "launcher": ""
+        },
+        "reference": "transpiler/common/errors/atomic/is_not_assignable.err"
+    }
+]
```
Comment (kchabSS) [inline tests/functional/configs/test_suite_transpiler/common/errors/atomic.json:62]: EOL at EOF

File: tests/functional/configs/test_suite_transpiler/common/errors/barrier.json [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+        "action_config": {
+            "backend": "cuda",
+            "source": "transpiler/common/errors/barrier/takes_at_most_one_argument.cpp",
+            "includes": [],
+            "defs": [],
+            "launcher": ""
+        },
+        "reference": "transpiler/common/errors/barrier/takes_at_most_one_argument.err"
+    }
+]
```
Comment (kchabSS) [inline tests/functional/configs/test_suite_transpiler/common/errors/barrier.json:62]: EOL at EOF

File: tests/functional/configs/test_suite_transpiler/common/errors/exclusive.json [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+        "action_config": {
+            "backend": "cuda",
+            "source": "transpiler/common/errors/exclusive/does_not_take_arguments.cpp",
+            "includes": [],
+            "defs": [],
+            "launcher": ""
+        },
+        "reference": "transpiler/common/errors/exclusive/does_not_take_arguments.err"
+    }
+]
```
Comment (kchabSS) [inline tests/functional/configs/test_suite_transpiler/common/errors/exclusive.json:50]: EOL at EOF

File: tests/functional/configs/test_suite_transpiler/common/errors/loops.json [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+        "action_config": {
+            "backend": "cuda",
+            "source": "transpiler/common/errors/loops/outer_inside_inner.cpp",
+            "includes": [],
+            "defs": [],
+            "launcher": ""
+        },
+        "reference": "transpiler/common/errors/loops/outer_inside_inner.err"
+    }
+]
```
Comment (kchabSS) [inline tests/functional/configs/test_suite_transpiler/common/errors/loops.json:98]: EOL at EOF

File: tests/functional/data/transpiler/backends/cuda/implicit/const_global_const_size_array.cpp [hunk 11]
```diff
@@ -11,3 +11,10 @@ const float arr_const3[] = {1., 2., 3., 4., 5., 6.};
 // Multidimensional
 const float arr_const4[][2] = {{1., 2.}, {3., 4.}, {5., 6.}};
 const float arr_const5[][3][2] = {{{1., 2.}, {3., 4.}, {5., 6.}}, {{1., 2.}, {3., 4.}, {5., 6.}}};
+
+// At least one @kern function is requried
+@kernel void kern () {
+    @outer for (int i = 0; i < 32; ++i) {
+        @inner for (int j = 0; j < 32; ++j) {}
+    }
+}
```
Comment (kchabSS) [inline tests/functional/data/transpiler/backends/cuda/implicit/const_global_const_size_array.cpp:20]: EOL at EOF

File: tests/functional/data/transpiler/backends/cuda/implicit/const_global_pointer.cpp [hunk 12]
```diff
@@ -11,3 +11,10 @@ int* const ptr_const4 = 0;
 
 // Stupid formatting
 const int* ptr_const5 = 0;
+
+// At least one @kern function is requried
+@kernel void kern () {
+    @outer for (int i = 0; i < 32; ++i) {
+        @inner for (int j = 0; j < 32; ++j) {}
+    }
+}
```
Comment (kchabSS) [inline tests/functional/data/transpiler/backends/cuda/implicit/const_global_pointer.cpp:20]: EOL at EOF

File: tests/functional/data/transpiler/backends/cuda/implicit/const_global_variable.cpp [hunk 13]
```diff
@@ -10,3 +10,10 @@ volatile int const var_const3 = 0;
 const int var_const4 = 0;
 
 int const var_const5 = 0;
+
+// At least one @kern function is requried
+@kernel void kern () {
+    @outer for (int i = 0; i < 32; ++i) {
+        @inner for (int j = 0; j < 32; ++j) {}
+    }
+}
```
Comment (kchabSS) [inline tests/functional/data/transpiler/backends/cuda/implicit/const_global_variable.cpp:19]: EOL at EOF

File: tests/functional/data/transpiler/backends/cuda/implicit/extern_const_global_array.cpp [hunk 14]
```diff
@@ -4,3 +4,10 @@ struct S {
 extern const int arr_0[];
 extern const float arr_1[];
 extern const S arr_2[];
+
+// At least one @kern function is requried
+@kernel void kern () {
+    @outer for (int i = 0; i < 32; ++i) {
+        @inner for (int j = 0; j < 32; ++j) {}
+    }
+}
```
Comment (kchabSS) [inline tests/functional/data/transpiler/backends/cuda/implicit/extern_const_global_array.cpp:13]: EOL at EOF

File: tests/functional/data/transpiler/backends/cuda/implicit/non_kernel_function.cpp [hunk 15]
```diff
@@ -5,3 +5,10 @@ static float add1(const float* a, int i, const float* b, int j) {
 float add2(const float* a, int i, const float* b, int j) {
     return a[i] + b[i];
 }
+
+// At least one @kern function is requried
+@kernel void kern () {
+    @outer for (int i = 0; i < 32; ++i) {
+        @inner for (int j = 0; j < 32; ++j) {}
+    }
+}
```
Comment (kchabSS) [inline tests/functional/data/transpiler/backends/cuda/implicit/non_kernel_function.cpp:14]: EOL at EOF

File: tests/functional/data/transpiler/backends/hip/implicit/const_global_const_size_array.cpp [hunk 16]
```diff
@@ -11,3 +11,10 @@ const float arr_const3[] = {1., 2., 3., 4., 5., 6.};
 // Multidimensional
 const float arr_const4[][2] = {{1., 2.}, {3., 4.}, {5., 6.}};
 const float arr_const5[][3][2] = {{{1., 2.}, {3., 4.}, {5., 6.}}, {{1., 2.}, {3., 4.}, {5., 6.}}};
+
+// At least one @kern function is requried
+@kernel void kern () {
+    @outer for (int i = 0; i < 32; ++i) {
+        @inner for (int j = 0; j < 32; ++j) {}
+    }
+}
```
Comment (kchabSS) [inline tests/functional/data/transpiler/backends/hip/implicit/const_global_const_size_array.cpp:20]: EOL at EOF

File: tests/functional/data/transpiler/backends/hip/implicit/const_global_variable.cpp [hunk 17]
```diff
@@ -10,3 +10,10 @@ volatile int const var_const3 = 0;
 const int var_const4 = 0;
 
 int const var_const5 = 0;
+
+// At least one @kern function is requried
+@kernel void kern () {
+    @outer for (int i = 0; i < 32; ++i) {
+        @inner for (int j = 0; j < 32; ++j) {}
+    }
+}
```
Comment (kchabSS) [inline tests/functional/data/transpiler/backends/hip/implicit/const_global_variable.cpp:19]: EOL at EOF

**Review by kchabSS — COMMENTED** [review]
Overall very nice! Some places need small adjustments.

File: lib/attributes/backend/dpcpp/exclusive.cpp [hunk 18]
```diff
@@ -1,8 +1,7 @@
 #include "attributes/attribute_names.h"
 #include "attributes/utils/cuda_subset/handle.h"
 #include "attributes/utils/default_handlers.h"
-#include "core/handler_manager/backend_handler.h"
-#include "core/utils/attributes.h"
+#include "core/handler_manager/handler_manager.h"
```
[the remark this answers is not in the document]
Comment (pavlo-hilei) [reply lib/attributes/backend/dpcpp/exclusive.cpp]: Done

File: lib/attributes/backend/dpcpp/kernel.cpp [hunk 19]
```diff
@@ -117,8 +117,8 @@ HandleResult handleKernelAttribute(SessionStage& s,
     auto typeStr = rewriter.getRewrittenText(func.getReturnTypeSourceRange());
     auto paramStr = getFunctionParamStr(func, oklKernelInfo, rewriter);
 
-    if (auto verified = verifyLoops(kernelInfo); !verified) {
-        return verified;
+    if (auto verified = verifyLoops(kernelInfo, s); !verified) {
```
[the remark this answers is not in the document]
Comment (pavlo-hilei) [reply lib/attributes/backend/dpcpp/kernel.cpp]: Done

File: lib/attributes/utils/common.cpp (hunk 2 above)
Comment (pavlo-hilei) [reply lib/attributes/utils/common.cpp]: Done

File: tests/functional/data/transpiler/backends/cuda/implicit/const_global_variable.cpp (hunk 13 above)
Comment (pavlo-hilei) [reply tests/functional/data/transpiler/backends/cuda/implicit/const_global_variable.cpp]: Done

File: lib/attributes/utils/serial_subset/exclusive.cpp [hunk 20]
```diff
@@ -1,3 +1,4 @@
+#include "attributes/utils/common.h"
```
[the remark this answers is not in the document]
Comment (pavlo-hilei) [reply lib/attributes/utils/serial_subset/exclusive.cpp]: Done

File: tests/functional/configs/test_suite_transpiler/backends/cuda/barrier.json (hunk 6 above)
Comment (pavlo-hilei) [reply tests/functional/configs/test_suite_transpiler/backends/cuda/barrier.json]: Done

File: tests/functional/configs/test_suite_transpiler/common/errors/barrier.json (hunk 8 above)
Comment (pavlo-hilei) [reply tests/functional/configs/test_suite_transpiler/common/errors/barrier.json]: Done

File: tests/functional/configs/test_suite_transpiler/common/errors/exclusive.json (hunk 9 above)
Comment (pavlo-hilei) [reply tests/functional/configs/test_suite_transpiler/common/errors/exclusive.json]: Done

File: tests/functional/configs/test_suite_transpiler/common/errors/loops.json (hunk 10 above)
Comment (pavlo-hilei) [reply tests/functional/configs/test_suite_transpiler/common/errors/loops.json]: Done

File: tests/functional/data/transpiler/backends/cuda/implicit/const_global_const_size_array.cpp (hunk 11 above)
Comment (pavlo-hilei) [reply tests/functional/data/transpiler/backends/cuda/implicit/const_global_const_size_array.cpp]: Done

File: tests/functional/data/transpiler/backends/cuda/implicit/const_global_pointer.cpp (hunk 12 above)
Comment (pavlo-hilei) [reply tests/functional/data/transpiler/backends/cuda/implicit/const_global_pointer.cpp]: Done

File: tests/functional/data/transpiler/backends/cuda/implicit/extern_const_global_array.cpp (hunk 14 above)
Comment (pavlo-hilei) [reply tests/functional/data/transpiler/backends/cuda/implicit/extern_const_global_array.cpp]: Done

File: tests/functional/data/transpiler/backends/cuda/implicit/non_kernel_function.cpp (hunk 15 above)
Comment (pavlo-hilei) [reply tests/functional/data/transpiler/backends/cuda/implicit/non_kernel_function.cpp]: Done

**Review by kchabSS — APPROVED** [review]
Blessing this PR.

**Pull request merged**